src/sentry/static/sentry/app/components/search/sources/routeSource.tsx

```diff
@@ -0,0 +1,185 @@
+import flattenDepth from 'lodash/flattenDepth';
+import React from 'react';
+import {FuseOptions, FuseResultWithMatches} from 'fuse.js';
+import {WithRouterProps} from 'react-router';
+
+import {createFuzzySearch} from 'app/utils/createFuzzySearch';
+import accountSettingsNavigation from 'app/views/settings/account/navigationConfiguration';
+import organizationSettingsNavigation from 'app/views/settings/organization/navigationConfiguration';
+import projectSettingsNavigation from 'app/views/settings/project/navigationConfiguration';
+import replaceRouterParams from 'app/utils/replaceRouterParams';
+import withLatestContext from 'app/utils/withLatestContext';
+import {NavigationItem} from 'app/views/settings/types';
+import {Organization, Project} from 'app/types';
+
+type Config =
+  | typeof accountSettingsNavigation
+  | typeof organizationSettingsNavigation
+  | typeof projectSettingsNavigation;
+
+// XXX(epurkhiser): We use the context in mapFunc to handle both producing the
+// NavigationSection list AND filtering out items in the sections that should
+// not be shown using the `show` attribute of the NavigationItem
+type Context = Parameters<Extract<Config, Function>>[0] &
+  Parameters<Extract<NavigationItem['show'], Function>>[0];
+
+/**
+ * navigation configuration can currently be either:
+ *
+ *  - an array of {name: string, items: Array<{NavItem}>} OR
+ *  - a function that returns the above
+ *    (some navigation items require additional context, e.g. a badge based on
+ *    a `project` property)
+ *
+ * We need to go through all navigation configurations and get a flattened list
+ * of all navigation item objects
+ */
+const mapFunc = (config: Config, context: Context | null = null) =>
+  (Array.isArray(config)
+    ? config
+    : context !== null
+    ? config(context)
+    : []
+  ).map(({items}) =>
+    items.filter(({show}) =>
+      typeof show === 'function' && context !== null ? show(context) : true
+    )
+  );
+
+type ItemProps = {
+  resultType: 'route';
+  sourceType: 'route';
+  to: string;
+};
+
+type Result = {
+  item: NavigationItem & ItemProps;
+  matches: FuseResultWithMatches<NavigationItem>['matches'];
+  score: number;
+};
+
+type RenderProps = {
```

Comment:
maybe `ChildrenProps` would be a better name?

```suggestion
type ChildrenProps = {
```

Reply:
We have a lot of usage of `RenderProps` for render-prop style components, so I think this is probably OK